docs/rfcs/017-timeline-data-management.md

```diff
@@ -0,0 +1,391 @@
+
+One of the solutions to the issue is already being developed for the remote storage sync: [SyncQueue](https://github.com/neondatabase/neon/blob/main/pageserver/src/storage_sync.rs#L463)
+The queue is able to batch CRUD layer operations (both for local and fs contexts) and reorder them to increase the sync speed.
+Similar approach could be generalized for all layer modifications, including in-memory ones such as GC or compaction: this way, we could manage all layer modifications and reads in one place with lesser locks and tests that are closer to unit tests.
+
+- change the approach to locking synchronization
+
+A number of locks in the timeline seem to be used to coordinate gc, compaction tasks and related processes.
+It should be done in a task manager or other place, external to the timeline.
+
+Timeline contents still needs to be synchronized, considering the task work, so fields like `latest_gc_cutoff_lsn: RwLock<Lsn>` are expected to stay for that purpose, but general amount of locks should be reduced.
+
+### Putting it all together
+
+If the proposal bullets applied to the diagrams above, the state could be represented as:
+
+![timeline timeline tenant state](./images/017-timeline-data-management/proposed_timeline_tenant_state.svg)
+
+The reorders aim to put all tasks into separated modules, with strictly defined interfaces and as less knowledge about other components, as possible.
+This way, all timeline data is now in the `data_storage`, including the GC, walreceiver, `RemoteTimelineIndex`, `LayerMap`, etc. with some API to get the data in the way,
+more conveninent for the data sync system inside.
+So far, it seems that a few maps with `Arc<RwLock<SeparateData>>` with actual data operations added inside each `SeparateData` struct, if needed.
+
+`page_cache` is proposed to placed into the same `data_storage` since it contains tenant timelines' data: this way, all metadata and data is in the same struct, simplifying things with Rust's borrow checker and allowing is to do care about data sync withing a single module.
+
+`task_manager` is related to data storage and manages all tenant and timeline tasks, manages shared resources (runtimes, threadpools, etcd connection, etc.) and synchronizes tasks.
+All locks such as `gc_cs` belong to this module tree, as primitives inherently related to the task synchronization.
+Tasks have to access timelines and their metadata, but should do that through `data_storage` API and similar.
+
+`task_manager` should (re)start, stop and track all tasks that are run in it, selecting an appropriate runtime depending on a task kind (we have async/sync task separation, CPU and IO bound tasks separation, ...)
+Some locks such as `layer_removal_cs` one are not needed, if the only component that starts the tasks ensures they don't run concurrently.
```

Comment:
Tasks can be started manually via an api. Is api expected to go to task_manager and ask it to run the task? Will the lock move inside task_manager then?

Reply:
I've considered the similar case with branching somewhere near, but yes, this is one odd place where the locks from `task_manager` have to be accessed in http handler.

No better ideas from me so far, but I don't think that those locks should be moved away from their task handling.

Reply:
> and share it between task manager and repo

Most probably, that what it would turn into, but at least initially, I wanted to consider the case where locks are taken explicitly from a single place: it's hard to track those `Arc<Mutex<..>>` usages across the code, harder to keep them locked in proper order, when the code and the number of locks grow (and that's what we try to solve now).

Both approaches seem to be very similar in essence, so unless we turn around the locks somehow, it does not matter which to pick, IMO.

Reply:
Yeah, so in this case with proposed task manager we need to share the lock between the task manager and the repo so it may become even more difficult to follow. We can write a comment that repository code does not lock them but only passes them to task manager which does that. Looks suboptimal to me, so  maybe we need to find a way to clean up existing locks inside repo and that would solve the problem. Maybe there are other alternatives...